Use incremental policy updates when identities are added or deleted.

jrajahalme opened the pull request:
Accumulate MapState entires due to identity changes and apply them
without full policy recalculation.

A policy update trigger is still needed like before, but the policy computation is skipped if not forced and the policy itself did not change.

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cilium/cilium/8263)
<!-- Reviewable:end -->

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -926,7 +927,7 @@ func (e *Endpoint) deletePolicyKey(keyToDelete policy.Key) error {
 	return nil
 }
 
-func (e *Endpoint) e.addPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntry) error {
```
Can you fix this in the `policy: Refactor policymap updates.` commit?

jrajahalme replied:
Done!

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -940,20 +941,57 @@ func (e *Endpoint) e.addPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntr
+
+	adds, deletes := e.realizedPolicy.PolicyMapChanges.GetMapChanges()
+
+	e.getLogger().Debugf("applyPolicyMapChanges: adds: %v, deletes: %v", adds, deletes)
```
Please use withfields here instead of printing the adds and deletes in the string.

jrajahalme replied:
Deleted this, was too noisy.

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -940,20 +941,57 @@ func (e *Endpoint) e.addPolicyKey(keyToAdd policy.Key, entry policy.MapStateEntr
+	if len(errors) > 0 {
+		return fmt.Errorf("updating desired PolicyMap state failed: %s", errors)
+	} else if len(adds)+len(deletes) > 0 {
+		e.getLogger().Infof("Applied policy updates due to added identities: %v, deleted identities: %v", adds, deletes)
```
Same here. Also, I don't think this needs to be `Infof`

jrajahalme replied:
Changed to debug with logfields.

aanm commented on pkg/policy/mapstate.go:
```diff
@@ -100,3 +101,74 @@ func (keys MapState) AllowAllIdentities(ingress, egress bool) {
+		ProxyPort: 0, // Will be updated by the caller when applicable
+	}
+
+	log.Debugf("MapChanges: AccumulateMapChanges(adds: %v, deletes: %v, port: %d, proto: %d, direction: %d)",
```
With fields.

jrajahalme replied:
Done!

aanm commented on pkg/policy/distillery.go:
```diff
@@ -117,10 +117,22 @@ func (cache *PolicyCache) updateSelectorPolicy(identity *identityPkg.Identity) (
-	if revision <= currentRevision {
+
+	// Don't resolve policy if it was already done for this or later revision.
+	if revision <= currentPolicy.Revision {
```
Why don't we get the `revision` after taking the lock of `cip`?

jrajahalme replied:
It is not necessary as the repo is read locked during this call, but I changed to this like you suggested here to make this clearer.

aanm commented on pkg/policy/l4.go:
```diff
@@ -148,6 +153,7 @@ func (l4 *L4Filter) ToKeys(direction trafficdirection.TrafficDirection) []Key {
 
 	if l4.AllowsAllAtL3() {
 		if l4.Port == 0 {
+			log.Debugf("ToKeys(%s): allow all", direction)
```
withfields

jrajahalme replied:
Done.

aanm commented on pkg/policy/l4.go:
```diff
@@ -157,6 +163,7 @@ func (l4 *L4Filter) ToKeys(direction trafficdirection.TrafficDirection) []Key {
 			keysToAdd = append(keysToAdd, keyToAdd)
 		} else {
 			// L4 allow
+			log.Debugf("ToKeys(%s): L4 allow all on %d/%d", direction, port, proto)
```
withfields

jrajahalme replied:
Done.

aanm commented on pkg/policy/l4.go:
```diff
@@ -463,7 +501,36 @@ func NewL4Policy(revision uint64) *L4Policy {
+func (l4 *L4Policy) insertUser(user *EndpointPolicy) {
+	l4.mutex.Lock()
+	if l4.users == nil {
+		log.Warningf("L4Filter: Skipping setting user on a stale policy")
```
This seems to be a Warning message, what can the user do if it seems this message?

jrajahalme replied:
Good question :-) Changed to Debug.

aanm commented on pkg/policy/l4.go:
```diff
@@ -463,7 +501,36 @@ func NewL4Policy(revision uint64) *L4Policy {
+func (l4 *L4Policy) insertUser(user *EndpointPolicy) {
+	l4.mutex.Lock()
+	if l4.users == nil {
+		log.Warningf("L4Filter: Skipping setting user on a stale policy")
```
`s/Warningf/Warning`

aanm commented on pkg/policy/l4.go:
```diff
@@ -463,7 +501,36 @@ func NewL4Policy(revision uint64) *L4Policy {
+	port uint16, proto uint8, direction trafficdirection.TrafficDirection) {
+
+	if l4 == nil {
+		log.Warning("L4Filter: Skipping incremental updates due to nil policy pointer")
```
Same here for the user inability of doing any action if it sees this message

jrajahalme replied:
Changed to Debug

nebril commented on pkg/policy/l4.go:
```diff
@@ -173,7 +180,7 @@ func (l4 *L4Filter) ToKeys(direction trafficdirection.TrafficDirection) []Key {
 	for _, cs := range l4.CachedSelectors {
 		identities := cs.GetSelections()
-		log.Debugf("ToKeys: Allowed remote IDs for selector %v: %v", cs, identities)
+		log.Debugf("ToKeys(%s): Allowed remote IDs for selector %v: %v", direction, cs, identities)
```
would be nice to use withfields here too

jrajahalme replied:
Done!

joestringer commented on pkg/policy/l4.go:
```diff
@@ -191,13 +214,27 @@ func (l4 *L4Filter) ToKeys(direction trafficdirection.TrafficDirection) []Key {
+		logfields.DeletedPolicyID:  deleted,
 	}).Debug("identities selected by L4Filter updated")
+
+	// Skip updates on filter that wildcards L3 and/or L4.
```
```suggestion
	// Skip updates on filter that wildcards L3.
```

This is only checking that L3 is wilcarded, right? Either `AllowsAllAtL3()` or there's no L3-dependent rules (meaning they're L7 rules with an L4 dependency only).

jrajahalme replied:
Correct, fixed comment as suggested.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -1048,24 +1122,19 @@ func (e *Endpoint) syncPolicyMap() error {
-			// converted to network byte-order.
-			err := e.PolicyMap.DeleteKey(keyHostOrder)
+		if _, ok := e.desiredPolicy.PolicyMapState[keyToDelete]; !ok {
+			e.getLogger().WithField(logfields.BPFMapKey, entry.Key.String()).Warning("syncPolicyMap removing a bpf policy entry not in the desired state")
```
I think we'll want to keep an eye on this for whether it ends up being really chatty in some adverse condition.

What can a user do if we warn on this?

jrajahalme replied:
If the new logic is correct this should not happen so it is useful to log these cases to know when this happens to help figure out why.

I'll change this to Debug to hide these in normal operation.

joestringer commented on pkg/policy/mapstate.go:
```diff
@@ -100,3 +104,79 @@ func (keys MapState) AllowAllIdentities(ingress, egress bool) {
+
+// GetMapChanges transfers the changes from MapChanges to the caller.
+// May return nil maps.
+func (mc *MapChanges) GetMapChanges() (adds, deletes MapState) {
```
```suggestion
func (mc *MapChanges) StealMapChanges() (adds, deletes MapState) {
```

or (wait for it)
```suggestion
func (mc *MapChanges) ConsumeMapChanges() (adds, deletes MapState) {
```
? 

This would be a bit more indicative that it's a write operation. That said, feel free to disagree.

ianvernon replied:
You guys know what naming convention I am for here ;)

jrajahalme replied:
I went with `ConsumeMapChanges()`, thanks for the suggestion :-)

joestringer commented on pkg/policy/distillery.go:
```diff
@@ -108,19 +108,27 @@ func (cache *PolicyCache) delete(identity *identityPkg.Identity) bool {
+	//
+	// Lock the 'cip' for the duration of the revision check and
+	// the possible policy update.
+	cip.Lock()
```
Not sure why this approach didn't occur to me before ;)

ianvernon replied:
Is it possible that this was the issue behind https://github.com/cilium/cilium/issues/8269 ? It seems quite related.

joestringer commented on pkg/policy/l4.go:
```diff
@@ -116,6 +119,12 @@ type L4Filter struct {
+	// This is set before concurrent access is possible, but set to nil without
+	// synchronization. This is only used to call AccumulateMapChanges(), which
+	// checks for the nil before dereferencing.
+	policy *L4Policy
```
If it's possible to be accessed concurrently, could we wrap it with an atomic type even for documentation purposes? IIRC there's a way to do this without any runtime cost on common architectures, right?

jrajahalme replied:
OK, will do. Pointer access is atomic on all CPU architectures, so there should be no run-time cost to it.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -903,48 +904,138 @@ func (e *Endpoint) GetBPFValue() (*lxcmap.EndpointInfo, error) {
 
 	if len(errors) > 0 {
-		return fmt.Errorf("deleting stale PolicyMap state failed: %s", errors)
+		return fmt.Errorf("updating desired PolicyMap state failed: %s", errors)
```
Given that the `addPolicyKey()` / `deletePolicyKey()` functions already log the individual errors, is it necessary to add them into this error message to return?

jrajahalme replied:
No, refactoring to remove the double logging.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -903,48 +904,138 @@ func (e *Endpoint) GetBPFValue() (*lxcmap.EndpointInfo, error) {
+}
+
+func (e *Endpoint) proxyID(key policy.Key) string {
+	return policy.ProxyID(e.ID, key.TrafficDirection == trafficdirection.Ingress.Uint8(), u8proto.U8proto(key.Nexthdr).String(), key.DestPort)
```
Can we fold this directly into `policy.ProxyID()`? I don't see why `pkg/endpoint` should be digging around using internals of a policy structure to pass to a function in the policy package.

jrajahalme replied:
Added new `policy.ProxyIDFromKey()` and `policy.ProxyIDFromFilter` to address these concerns.

joestringer commented on pkg/endpoint/bpf.go:
```diff
@@ -903,48 +904,138 @@ func (e *Endpoint) GetBPFValue() (*lxcmap.EndpointInfo, error) {
+			}
+		}
+
+		err := e.addPolicyMapDelta()
```
I'm a bit confused, we now have `addPolicyMapDelta()`, `syncPolicyMapDelta()` and `applyPolicyMapChanges()`.. surely we don't need three such similarly-named functions? I think there's some distinctions here between the roles of each of these, but the naming is really not helping.

jrajahalme replied:
`sync` consists of adding and deleting, so `syncPolicyMapDelta()` calls `addPolicyMapDelta()`. These generate the `delta` by comparing maps. A sync without the "Delta" dumps the bpf map for the basis of comparison.

`applyPolicyMapChanges()` applies changes/deltas without the need for generating the delta by exhaustively comparing the maps.

All these functions have a different function, so we need them. trying to come up with better naming, but drawing a blank so far.

jrajahalme replied:
Renaming `syncPolicyMap()` as `syncPolicyMapWithDump()` and `syncPolicyMapDelta()` as `syncPolicyMap()`. This gives a clearer indication which of them carries the cost of a full bpf map dump, and avoids using the `Delta` suffix that has a connotation with the incremental policy computation feature, while the function does a full diff between two maps.

joestringer reviewed (changes requested):
Just a few pretty minor code organization / readability comments below, but looks pretty good---and the benchmarks I did earlier look pretty positive too. :+1:

aanm commented on pkg/endpoint/bpf.go:
```diff
@@ -903,48 +904,138 @@ func (e *Endpoint) GetBPFValue() (*lxcmap.EndpointInfo, error) {
+func (e *Endpoint) syncPolicyMapDelta() error {
+	// Nothing to do if the desired policy is already fully realized.
+	if e.realizedPolicy != e.desiredPolicy {
+		errors := []error{}
```
`var errors []error`

jrajahalme replied:
This has now been replaced by `errors := 0`

ianvernon commented on daemon/policy.go:
```diff
@@ -73,7 +73,7 @@ func (d *Daemon) TriggerPolicyUpdates(force bool, reason string) {
 func (d *Daemon) UpdateIdentities(added, deleted cache.IdentityCache) {
 	d.policy.GetSelectorCache().UpdateIdentities(added, deleted)
-	d.TriggerPolicyUpdates(true, "one or more identities created or deleted")
+	d.TriggerPolicyUpdates(false, "one or more identities created or deleted")
```
Great to see us removing more calls to forcibly regenerate all endpoints!

ianvernon commented on pkg/policy/mapstate.go:
```diff
@@ -100,3 +101,74 @@ func (keys MapState) AllowAllIdentities(ingress, egress bool) {
+		adds, deletes, port, proto, direction.Uint8())
+
+	mc.mutex.Lock()
+	if len(adds) > 0 {
```
Does the ordering of how we do processing of additions / deletions matter here? It seems like if we have an identity in `adds` but also in `deletes`, that we'd end up with it deleted, but if we processed it in the inverse order (deletes and then adds), we'd have it in `adds`, but not `deletes`.

jrajahalme replied:
The sets 'adds' and 'deletes' are disjoint so that no identity is present in both. I added a commit documenting this and verified that this is the case on all call sites.

ianvernon commented on pkg/policy/l4.go:
```diff
@@ -463,7 +503,36 @@ func NewL4Policy(revision uint64) *L4Policy {
+	port uint16, proto uint8, direction trafficdirection.TrafficDirection) {
+
+	if l4 == nil {
+		log.Debug("L4Filter: Skipping incremental updates on stale (detached) policy.")
```
There is no unique identifying information regarding which `L4Policy` this is coming from - so the debugging logs won't be able to correlate the associated endpoint with which this `L4Policy` is associated.


The log message should also say `L4Policy` instead of `L4Filter`.

jrajahalme replied:
Removed these logs.

ianvernon reviewed (commented):
This is great overall. I have a few questions / miscellaneous comments.

jrajahalme committed:
policy: Pass policy revision to NewL4Policy().

Policy repository revision is known at the time when the L4Policy is
created, so pass it to the constructor rather than setting it directly
after.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
u8proto: Be compatible with policy/api

Change the string form of the wildcard protocol from "All" to "ANY" to
be compatible with the definitions in policy/api.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Simplify syntax

Simplify policy resolution syntax by passing L4PolicyMap{} instead of
a full L4Policy when not needed, and by returning a map instead of a
pointer.

This avoids multiple calls to NewL4Policy(0) with a fake revision 0,
when only the L4PolicyMap for ingress or egress is being tested.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Refactor policymap updates.

Policy map updates will be done also incrementally later, refactor to
facilitate.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Log policy map sync deletes

The full policy map sync is only done from the controller. Gain some
visibility to any discrepencies in this controller sync in order to
gain visibility to potenial transitent policy map errors.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Remove dead testing code.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Protect against racing policy updates.

Racing policy updates for the same identity were harmless before, but
the need to maintain backpointers to the endpoint policies using the
selector policy made this more complicated. To keep the selector
policy life cycle as simple as possible, apply mutual exclusion on
updates on the policy of the same identity. This also saves some CPU.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
daemon: Do not bump policy revision on identity changes.

Identity changes are noted via a separate revision number maintained
by the selector cache, so it is no longer necessary to artificially
bump the policy revision number when identities have changed.

Add unit testing code that exercises the identity change triggering
code that is changed to make sure policy computation still works.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Fix logging

Use log fields.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Introduce MapChanges

MapChanges stores MapState maps that can accumulate additions and
deletions of policy map entries. This will be used for incremental
policy updates.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Accumulate MapChanges for identity changes

Accumulate MapState entires due to identity changes.

This requires the introduction of backpointers from each L4Filter back
to the L4Policy, and from the L4Policy to the EndpointPolicies that
are using them. This way the L4Filters can update the MapChanges
maintained by the EndpointPolicies. The backpointers required for this
make these data structures circular. Thus, these pointers must be
cleaned up so that GC can reclaim the memory.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Clarify locking.

Take the repo revision after acquiring the lock on the cached
policy. This does not change anything as the repo is read locked, but
makes this more obviously correct.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Clarify syncPolicyMap function naming

Rename syncPolicyMap() as syncPolicyMapWithDump() and
syncPolicyMapDelta() as syncPolicyMap(). This gives a clearer
indication which of them carries the cost of bpf map dump, and avoids
using the "Delta" suffix that has a connotation to the planned
incremental policy update feature that is not yet implemented.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
endpoint: Use accumulated map changes for policy updates

Use the accumulated map changes for bpf policy map updates when the
policy remains the same, but identities have changed.

This change removes the need for the endpoint to track the last
applied identity revision number.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Reduce logging.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Require identity adds, deletes be disjoint.

Document that the identities being added and deleted are two disjoint
sets when passing them in identity updates. This allows the sets to be
processed in either order.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
policy: Fix MatchPattern formatting

Add the missing colon in the string representation.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

jrajahalme committed:
daemon: Do not force policy regeneration on FQDN changes

Now that FQDN selector changes are encoded in MapChanges we do not
need to force policy recalculation of all the endpoints by bumping the
repo revision.

Fix the runtime accordingly.

Signed-off-by: Jarno Rajahalme <jarno@covalent.io>

ianvernon commented on daemon/fqdn.go:
```diff
@@ -124,7 +124,7 @@ func (d *Daemon) updateSelectorCacheFQDNs(selectors map[policyApi.FQDNSelector][
 		return
 	}
-	d.TriggerPolicyUpdates(true, "updated identities for FQDNs")
+	d.TriggerPolicyUpdates(false, "updated identities for FQDNs")
```
<3

ianvernon reviewed (approved):
Ship it!
